Zend/zend_execute.c

```diff
@@ -673,70 +656,15 @@ static ZEND_COLD void zend_verify_type_error_common(
 		*fclass = "";
 	}
 
-	if (is_union_type(arg_info->type)) {
-		zend_string *type_str = zend_type_to_string(arg_info->type);
-		smart_str_appends(&str, "be of type ");
-		smart_str_append(&str, type_str);
-		zend_string_release(type_str);
-	} else if (ZEND_TYPE_HAS_CLASS(arg_info->type)) {
-		zend_bool is_interface = 0;
-		zend_class_entry *ce = *cache_slot;
-		if (!ce) {
-			ce = zend_fetch_class(ZEND_TYPE_NAME(arg_info->type),
-				(ZEND_FETCH_CLASS_AUTO | ZEND_FETCH_CLASS_NO_AUTOLOAD));
-		}
-		if (ce) {
-			if (ce->ce_flags & ZEND_ACC_INTERFACE) {
-				smart_str_appends(&str, "implement interface ");
-				is_interface = 1;
-			} else {
-				smart_str_appends(&str, "be an instance of ");
-			}
-			smart_str_append(&str, ce->name);
-		} else {
-			/* We don't know whether it's a class or interface, assume it's a class */
-			smart_str_appends(&str, "be an instance of ");
-			smart_str_append(&str, ZEND_TYPE_NAME(arg_info->type));
-		}
-
-		if (ZEND_TYPE_ALLOW_NULL(arg_info->type)) {
-			smart_str_appends(&str, is_interface ? " or be null" : " or null");
-		}
-	} else {
-		uint32_t type_mask = ZEND_TYPE_PURE_MASK_WITHOUT_NULL(arg_info->type);
-		switch (type_mask) {
-			case MAY_BE_OBJECT:
-				smart_str_appends(&str, "be an object");
-				break;
-			case MAY_BE_CALLABLE:
-				smart_str_appends(&str, "be callable");
-				break;
-			case MAY_BE_ITERABLE:
-				smart_str_appends(&str, "be iterable");
-				break;
-			default:
-			{
-				/* Hack to print the type without null */
-				zend_type type = arg_info->type;
-				ZEND_TYPE_FULL_MASK(type) &= ~MAY_BE_NULL;
-				zend_string *type_str = zend_type_to_string(type);
-				smart_str_appends(&str, "be of the type ");
-				smart_str_append(&str, type_str);
-				zend_string_release(type_str);
-				break;
-			}
-		}
-
-		if (ZEND_TYPE_ALLOW_NULL(arg_info->type)) {
-			smart_str_appends(&str, " or null");
-		}
-	}
+	zend_string *type_str = zend_type_to_string_resolved(arg_info->type, zf->common.scope);
+	smart_str_append(&str, type_str);
+	zend_string_release(type_str);
```

Comment:
No point in going through a smart_str here anymore, you can just do `*need_msg = zend_type_to_string_resolved(...)`